### PR TITLE
PO feedback on users WS

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 new issue:
* 1 minor

Details:
* Add a new line at the end of this file. (org.codehaus.sonar:sonar-core:src/main/java/org/sonar/core/user/UserGroupCount.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

File: server/sonar-server/src/main/java/org/sonar/server/user/UserUpdater.java [hunk 1]
```diff
@@ -65,13 +63,15 @@
   private final DbClient dbClient;
   private final UserIndexer userIndexer;
   private final System2 system2;
+  private final SecurityRealmFactory realmFactory;
 
-  public UserUpdater(NewUserNotifier newUserNotifier, Settings settings, DbClient dbClient, UserIndexer userIndexer, System2 system2) {
+  public UserUpdater(NewUserNotifier newUserNotifier, Settings settings, DbClient dbClient, UserIndexer userIndexer, System2 system2, SecurityRealmFactory realmFactory) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserUpdater.java]: what is this SecurityRealmFactory ?

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java [hunk 2]
```diff
@@ -67,18 +81,34 @@ public void define(WebService.NewController controller) {
...
   @Override
   public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+
     SearchOptions options = new SearchOptions()
       .setPage(request.mandatoryParamAsInt(Param.PAGE), request.mandatoryParamAsInt(Param.PAGE_SIZE));
     List<String> fields = request.paramAsStrings(Param.FIELDS);
     SearchResult<UserDoc> result = userIndex.search(request.param(Param.TEXT_QUERY), options);
 
+    Multimap<String, String> groupsByLogin = Multimaps.forMap(Maps.<String, String>newHashMap());
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java]: why use this constructor specifically ?

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java [hunk 3]
```diff
@@ -67,18 +81,34 @@ public void define(WebService.NewController controller) {
...
+    try {
+      Collection<String> logins = Collections2.transform(result.getDocs(), new Function<UserDoc, String>() {
+        @Override
+        public String apply(@Nonnull UserDoc input) {
+          return input.login();
+        }
+      });
+      groupsByLogin = dbClient.groupMembershipDao().selectGroupsByLogins(session, logins);
+    } finally {
+      session.close();
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java]: MyBatis.closeQuietly(session) ?

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java]: rename session in dbSession ?

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java [hunk 4]
```diff
@@ -86,6 +116,7 @@ private void writeUsers(JsonWriter json, SearchResult<UserDoc> result, @Nullable
       writeIfNeeded(json, user.login(), FIELD_LOGIN, fields);
       writeIfNeeded(json, user.name(), FIELD_NAME, fields);
       writeIfNeeded(json, user.email(), FIELD_EMAIL, fields);
+      writeGroupsIfNeeded(json, groupsByLogin.get(user.login()), fields);
       if (fieldIsWanted(FIELD_SCM_ACCOUNTS, fields)) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java]: extract in a writeScmAccountsIfNeeded to have the same level of abstraction ?

File: server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java [hunk 5]
```diff
@@ -79,6 +78,9 @@
   @Mock
   NewUserNotifier newUserNotifier;
 
+  @Mock
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java]: could we remove the MockitoJUnitRunner dependency ?

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/ChangePasswordActionTest.java [hunk 6]
```diff
@@ -51,7 +56,9 @@
 import static com.google.common.collect.Lists.newArrayList;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
+@RunWith(MockitoJUnitRunner.class)
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/ChangePasswordActionTest.java]: I'd prefer no to use the Mockito runner avec declare explicitly Mockito.mock(…) the mocks. I think it's the direction we're taking on the platform

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/SearchActionTest.java [hunk 7]
```diff
@@ -20,44 +20,84 @@
...
 import org.sonar.server.user.index.UserDoc;
 import org.sonar.server.user.index.UserIndex;
 import org.sonar.server.user.index.UserIndexDefinition;
 import org.sonar.server.ws.WsTester;
 
-import java.util.Arrays;
-
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class SearchActionTest {
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/SearchActionTest.java]: please add @Category(DbTests.class)

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/CreateActionTest.java [hunk 8]
```diff
@@ -52,8 +54,6 @@
 import org.sonar.server.user.index.UserIndexer;
 import org.sonar.server.ws.WsTester;
 
-import java.util.Locale;
-
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/CreateActionTest.java]: add @Category(DbTests.class) ?

Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/CreateActionTest.java]: remove use MockitoJUnitRunner.class ?

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java [hunk 9]
```diff
@@ -43,6 +43,7 @@
 import org.sonar.server.exceptions.NotFoundException;
 import org.sonar.server.tester.UserSessionRule;
 import org.sonar.server.user.NewUserNotifier;
+import org.sonar.server.user.SecurityRealmFactory;
 import org.sonar.server.user.UserUpdater;
 import org.sonar.server.user.db.UserDao;
 import org.sonar.server.user.index.UserDoc;
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java]: remove MockitoJUnitRunner dependency ?

Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java]: add @Category(DbTests.class)

File: sonar-core/src/main/java/org/sonar/core/user/GroupMembershipDao.java [hunk 10]
```diff
@@ -58,6 +63,22 @@ public int countGroups(SqlSession session, GroupMembershipQuery query, Long user
     return mapper(session).countGroups(params);
   }
 
+  public Multimap<String, String> selectGroupsByLogins(final DbSession session, Collection<String> logins) {
+    final Multimap<String, String> result = ArrayListMultimap.create();
+    DaoUtils.executeLargeInputs(logins, new NonNullInputFunction<List<String>, List<LoginGroup>>() {
```
Comment (teryk) [inline sonar-core/src/main/java/org/sonar/core/user/GroupMembershipDao.java]: we should not use NonNullInputFunction anymore, if my understanding is correct but the @NonNull annotation to override the Function contract

File: sonar-core/src/main/resources/org/sonar/core/user/GroupMembershipMapper.xml [hunk 11]
```diff
@@ -32,4 +32,17 @@
     <include refid="commonClauses" />
   </select>
 
+  <select id="selectGroupsByLogins" parameterType="string" resultType="org.sonar.core.user.LoginGroup">
+    SELECT u.login as login, g.name as groupName
+    FROM users u
+    LEFT JOIN groups_users gu ON gu.user_id=u.id
+    INNER JOIN groups g ON gu.group_id=g.id
+    <where>
```
Comment (teryk) [inline sonar-core/src/main/resources/org/sonar/core/user/GroupMembershipMapper.xml]: what's the advantage to use the MyBatis where clause ?

File: server/sonar-server/src/main/java/org/sonar/server/user/UserUpdater.java (hunk 1 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserUpdater.java]: As its name implies, it provides access to the SecurityRealm implementation used by the platform, if any (LDAP, Crowd, etc.)

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java (hunk 2 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java]: I don't care about the returned implementation. I could have used e.g ArrayListMultimap.create().

File: server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java (hunk 5 above)
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java]: Not here, see the ``NewUserNotifier`` stuff above - the object graph below is just too heavy. Besides, we don't want to couple this WS with the implementation of ``SecurityRealmFactory``.

Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java]: I didn't say to not mock, just don't use annotation with the Mockito runner. Declare mocks using Mockito.mock(…) method

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java (hunk 2 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/SearchAction.java]: if you don't mind, please do, simpler to read

File: sonar-core/src/main/resources/org/sonar/core/user/GroupMembershipMapper.xml (hunk 11 above)
Comment (jblievremont) [inline sonar-core/src/main/resources/org/sonar/core/user/GroupMembershipMapper.xml]: IIRC, the ``<select/>`` tag trims empty conditions.

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

**Pull request closed without merging**